### ISSUE TITLE
Function aliases: stdout/stderr not flushed, sys.exit not handled

### ISSUE THREAD

**Issue by blahgeek** [xonsh/xonsh#1819]
On current master (seems related to #1758 ) function aliases is pretty broken... e.g.

```
def _f(args, stdin=None):
    import time
    for i in range(10):
        time.sleep(1)
        print(i)

aliases['f'] = _f
f
```
This does not prints numbers one by one.

and:

```
def _f(args, stdin=None):
    import sys
    sys.exit(0)

aliases['f'] = _f
f
````
This does not terminate

Both example works properly on f48baa7

**Comment by astronouth7303** [comment]
I think for the first example, you should use the [streaming arguments](http://xon.sh/tutorial.html#callable-aliases). (I don't know if it'll flush any better, though. I think you're relying on Python's flushing behavior.)

In general xonsh does not currently patch the system values (`sys.stdin`, `os.environ`, `sys.exit()`, etc), especially per-execution.

**Comment by blahgeek** [comment]
I also tried streaming arguments, using `print('hello', file=stderr, flush=True)`, there's no difference. 

Both example works several days ago at f48baa7, so I think these should be bugs introduced by the new subproc execution model.

**Comment by blahgeek** [comment]
Update: if the function produces lots of output, the thread would stuck:

```
BlahGeeks-MBP ~/.local> def _f(args, stdin=None):
.......................     for i in range(1000):
.......................         print('x' * 100)
.......................         
BlahGeeks-MBP ~/.local> aliases['f'] = _f
BlahGeeks-MBP ~/.local> f
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
<stuck here>
```

**Comment by scopatz** [comment]
Ok, I think this is because the ProcProxy classes don't do non-blocking reads. We have the tools to fix this.

**Comment by scopatz** [comment]
Ok, can you try out #1870, now please?  That branch fixes the remaining issues here for me.

**Comment by blahgeek** [comment]
That works great for me, thanks a million!

**Comment by scopatz** [comment]
Great! hopefully that will get merged soon then

**Comment by gforsyth** [comment]
Fixed in #1870
